### PR TITLE
[9.2] Add clipboard button to public share link

### PR DESCRIPTION
**Pull request by DeepDiver1975**
bildschirmfoto von 2016-07-08 14-30-50

and I also moved the mail send button:
bildschirmfoto von 2016-07-08 15-22-52

### CODE REVIEW

**Comment by SergioBertolinSG** [comment]
screen shot 2016-07-08 at 14 51 10

Icon appears before sharing by link. It copies to clipboard "use-clipboard-api"

**Comment by SergioBertolinSG** [comment]
In safari, the copied string is the path of the file, not the public shared link.

Something like host/index.php/apps/files/?dir=/COSAS/SUBCOSAS&fileid=10
Instead of  host/index.php/s/0wmlnsYFkKh5fpT

**Comment by DeepDiver1975** [comment]
> Icon appears before sharing by link. It copies to clipboard "use-clipboard-api"

fixed

**Comment by DeepDiver1975** [comment]
> In safari, the copied string is the path of the file, not the public shared link.
> 
> Something like host/index.php/apps/files/?dir=/COSAS/SUBCOSAS&fileid=10
> Instead of host/index.php/s/0wmlnsYFkKh5fpT

strange :hankey:

**Comment by jancborchardt** [comment]
Ideally the link should directly be copied to the clipboard as soon as you click »Share link«. That will make the flow even more seamless. Can you change it so that happens automatically?

**Comment by jvillafanez** [comment]
@jancborchardt I haven't seen that behaviour in any other app. It seems unnatural that clicking in the checkbox copies the text. What if the user unchecks it later? What would happen with the clipboard?

In addition, users won't know that the link is copied unless they're told explicitly. If we don't use any kind of notification, this feature won't be used properly. So, where and how will the notification be placed in this case?

Moreover, the previous clipboard contents will be overwritten. I'd rather have an explicit user interaction. We don't know what will the user do with the link, and we shouldn't suppose that he will want to copy the link.

**Comment by DeepDiver1975** [comment]
> @jancborchardt I haven't seen that behaviour in any other app. It seems unnatural that clicking in the checkbox copies the text. What if the user unchecks it later? What would happen with the clipboard?
> 
> In addition, users won't know that the link is copied unless they're told explicitly. If we don't use any kind of notification, this feature won't be used properly. So, where and how will the notification be placed in this case?
> 
> Moreover, the previous clipboard contents will be overwritten. I'd rather have an explicit user interaction. We don't know what will the user do with the link, and we shouldn't suppose that he will want to copy the link.

Notifications can be implemented using the tooltip - this should not be an issue.

But I agree on the matter that the clipboard is overwritten which is somewhat unexpected and make the user angry if some valuable content is in there.

**Comment by MorrisJobke** [comment]
> we shouldn't suppose that he will want to copy the link.

But creating a link without copying it is kind of senseless, isn't it? I never created a link share (beside testing of the sharing app) and didn't copied it.

**Comment by MorrisJobke** [comment]
> But creating a link without copying it is kind of senseless, isn't it? I never created a link share (beside testing of the sharing app) and didn't copied it.

I just have no thing in mind: sending it via email: but then there should be definitely something like a "guide": 

you created a share: What to do with this share? Send via email - copy link - ...

**Comment by DeepDiver1975** [comment]
> I just have no thing in mind: sending it via email: but then there should be definitely something like a "guide":
> 
> you created a share: What to do with this share? Send via email - copy link - ...

There are for sure things which can be improved - let's for now just introduce the copy link button.
There is nothing really playing against this - right?

**Comment by jancborchardt** [comment]
Yeah, let’s go step-by-step. I still think automatically copying it makes the flow even easier and reduces a step you are _surely_ gonna take, which is copying the link.

The clipboard button we need in any case for when you want to get an existing shared link quickly, so :+1:

**Comment by DeepDiver1975** [comment]
@SergioBertolinSG safari has been fixed as far as technically possible.

This is the way this works once the user with safari hits the button: 
- the text in the input field is selected
- a tool tip will popup telling the user to press cmd+C to copy

In addition on ipad and ihone the tooltiop will tell the user that it is not supported - maybe better in this case would be to simply not show the button - but this is a future improvment from my pov

**Comment by SergioBertolinSG** [comment]
Works fine 👍 

Tested on chrome, safari (with that workaround) and edge.

File: core/js/sharedialoglinkshareview.js [hunk 1]
```diff
@@ -112,6 +115,38 @@
...
+				$input.tooltip({placement: 'bottom', trigger: 'manual', title: t('core', 'Copied!')});
+				$input.tooltip('show');
+				_.delay(function() {
+					$input.tooltip('hide');
+				}, 3000);
+			});
+			clipboard.on('error', function (e) {
+				$input = $(e.trigger);
+				var actionMsg = '';
+				if (/iPhone|iPad/i.test(navigator.userAgent)) {
```
Comment (PVince81) [inline core/js/sharedialoglinkshareview.js]: if we ever want to reuse that button elsewhere, would be good to have all this logic in a central place, like js.js in OC.Util or so

Comment (DeepDiver1975) [inline core/js/sharedialoglinkshareview.js]: We can do this once necessary - I have no clue what the perf impact is if we setup the clipboard instance in js.js

Comment (PVince81) [inline core/js/sharedialoglinkshareview.js]: Agreed. Note that the way I see it is more like a method `OC.Util.setupClipboard($inputField)` that does this lazily.

Anyway, let's merge this for now.

Comment (DeepDiver1975) [inline core/js/sharedialoglinkshareview.js]: > Anyway, let's merge this for now.

THX :tada:

**Comment by fcturner** [comment]
Great, guys! For my [Passwords app](https://GitHub.com/fcturner/passwords) I really would like to have it in a global function. 
For now, I'll delete my own dependency and copy your code. Thanks again!

**Comment by DeepDiver1975** [comment]
@fcturner please have a look at #25757 - it ships the global function - let me know if this works out for you - THX

**Comment by fcturner** [comment]
That's awesome! I'll let you know, thanks!

**Pull request merged**